Add a versus_test method to TestPlayer

drvinceknight opened the pull request:
Addresses #874 with a tweaked testing framework.

The `versus_test` method removes the option to set player histories and can be used to test strategies (and player attributes during matches) by creating a match between the player in question and an opponent.

The opponent can either be an actual player from the library or is defined as a cycle by a passed sequence of actions (there are examples of this in the docs and `test_titfortat.py`).

I've refactored all the tests in `test_titfortat.py` (which is now one module that raises no warnings).

When/if we're happy with this as a suggestion, we should open an issue about refactoring all the tests to use this approach. (This would be a big/important piece of work so we should ask for small contributions at a time to make sure we don't miss anything with the reviews).

Once that refactor is complete we could potentially remove unused methods in the `TestPlayer` class.

marcharper commented:
I like most of this. Some quick thoughts:

* Having opponent be a player or a sequence is a bit confusing -- using a MockPlayer would be better IMO, that's its purpose.
* We might be able to get away with only having init_kwargs, or it may be better to tackle that issue before we rewrite a bunch of tests.

drvinceknight commented:
> Having opponent be a player or a sequence is a bit confusing -- using a MockPlayer would be better IMO, that's its purpose.

Yeah I wasn't completely sure about having two input possibilities. I'll make this tweak.

> We might be able to get away with only having init_kwargs, or it may be better to tackle that issue before we rewrite a bunch of tests.

Can you expand on this a bit more? I'm not too sure I follow which issue you mean.

drvinceknight commented:
I've also just pushed a refactor of the `Random` strategy. I've used `init_kwargs` there.

marcharper commented:
Issue #706 --  maybe we could make all the Player args into kwargs, which I think would help with #706 as well.

drvinceknight commented:
> Issue #706 -- maybe we could make all the Player args into kwargs, which I think would help with #706 as well.

Sure thing, I'll remove the `init_args` option for now so that it will push us in that direction.

marcharper commented:
It's ok to use `test_reset`.

drvinceknight commented:
> It's ok to use test_reset.

Did you mean to type that here or in the other issue? Surely that's overwriting the `test_reset` method of the parent class?

marcharper commented on axelrod/tests/unit/test_grumpy.py:
```diff
@@ -45,7 +45,7 @@ def test_strategy(self):
                                          "nice_threshold": 0})
 
-    def test_reset(self):
+    def test_reset_state(self):
```
I think this test is unnecessary.

marcharper commented on axelrod/tests/unit/test_player.py:
```diff
@@ -288,8 +288,61 @@ def second_play_test(self, rCC, rCD, rDC, rDD, seed=None):
+            actions is passed, a Mock Player is created that cycles over that
+            sequence.
+        expected_outcomes: List
+            the expected outcomes of the match (list of tuples of actions).
```
T

marcharper commented on axelrod/tests/unit/test_player.py:
```diff
@@ -288,8 +288,61 @@ def second_play_test(self, rCC, rCD, rDC, rDD, seed=None):
+            `{length:-1}` implies that the players do not know the length of the
+            match.
+        attrs: dict
+            dictionary of internal attributes to check at the end of all plays
```
D

marcharper commented on axelrod/tests/unit/test_rand.py:
```diff
@@ -24,7 +24,24 @@ def test_strategy(self):
-        self.responses_test([C], [C, D, C], [C, C, D], seed=1)
+
+        opponent = axelrod.MockPlayer()
+        outcomes = [(C, C), (D, C), (D, C)]
```
This tests the histories above but not the next move, so it's not exactly equivalent. Doesn't matter in this case of course.

marcharper commented on axelrod/tests/unit/test_titfortat.py:
```diff
@@ -30,12 +30,42 @@ class TestTitForTat(TestPlayer):
+        self.second_play_test(rCC=C, rCD=D, rDC=C, rDD=D)
+
+        # Play against opponents
+        outcomes = [(C, C), (C, D), (D, C), (C, D)]
```
Same comment as before re: not checking the last move anymore.

drvinceknight commented:
@marcharper the particular test in `test_grumpy` was not covered by the default case as it's checking a non default value **but** your comment got me to find a bug in our default test case. It was in effect not testing anything (it was simply comparing the attributes of the clone to the attributes of the clone and not the reset player).

I've fixed this (cff2e53 and a refactor here: b7ce33a) and ended up having to fix some strategies that were then failing. 

This is worth a slow and thorough review in case I've missed anything so no rush :)

marcharper commented on axelrod/tests/unit/test_human.py:
```diff
@@ -91,3 +91,7 @@ def test_strategy(self):
         actual_action = human.strategy(Player(), lambda: C)
         self.assertEqual(actual_action, expected_action)
+
+    def test_reset_history_and_attributes(self):
```
Is this necessary? I though there was a catch in TestPlayer for this strategy. If you've removed that then ignore.

marcharper commented on axelrod/tests/unit/test_player.py:
```diff
@@ -207,37 +208,48 @@ def test_match_attributes(self):
-        for k, v in clone.__dict__.items():
-            if isinstance(v, np.ndarray):
-                self.assertTrue(np.array_equal(v, getattr(clone, k)))
+        for attribute, value in player.__dict__.items():
```
Would `player == clone` work here now?

marcharper commented on axelrod/tests/unit/test_meta.py:
```diff
@@ -46,16 +46,21 @@ def classifier_test(self, expected_class_classifier=None):
+        player.play(opponent)
+        player.reset()
+        for i, p in enumerate(player.team):
+            self.assertEqual(len(p.history), 0)
```
Can we use `player == clone` here?

marcharper commented on axelrod/tests/unit/test_player.py:
```diff
@@ -207,37 +208,48 @@ def test_match_attributes(self):
+            axelrod.seed(seed)
+            player.play(opponent)
+
+        player.reset()
```
`player == clone`?

marcharper commented:
That's interesting. I remember catching a subtle bug with MetaPlayer with that test so that's surprising that it worked at all! I like the new additions, looks like maybe we could remove some redundant code by using `player == clone` in a few places.

We might consider updating MetaPlayer's clone method to clone all the team members, or just take a closer look at it.

drvinceknight commented:
> That's interesting. I remember catching a subtle bug with MetaPlayer with that test so that's surprising that it worked at all! I like the new additions, looks like maybe we could remove some redundant code by using player == clone in a few places.

You might be able to tell from the commit history that I implemented a player `__eq__` but felt that that was stepping a bit too far for this PR so I reverted it (we should squash these commits for with the merge). Without the `__eq__` method `player == clone` would just look at instance equality (so be equivalent to `player is clone`). That could be added at another date.

> 
> We might consider updating MetaPlayer's clone method to clone all the team members, or just take a closer look at it.

Perhaps for another PR?

drvinceknight commented:
I've just rebased on to `master` and caught a misclassification of `stalker` :)

I'm not sure if this means I'm going to miss any comments you wrote but I think all of them were suggesting `player == clone` which (unless I'm missing something) won't work because of the fact that we haven't got a `__eq__` method implemented but that could be another issue to raise/discuss?

marcharper commented:
OK. `__eq__` was still there when I was commenting. I would vote to add it back in, but it can be another PR.

marcharper commented on axelrod/strategies/finite_state_machines.py:
```diff
@@ -57,6 +64,7 @@ def __init__(self, transitions=None, initial_state=None,
             initial_action = C
         super().__init__()
         self.initial_state = initial_state
+        self.state = initial_state
```
What's the purpose of this line?

drvinceknight replied:
Needed for the reset test. The cloned player did not have a `self.state` attribute if it had not played yet.

marcharper commented:
Do we want to change `expected_outcomes` to `expected_actions` or something else more inline with the terminology elsewhere in the library?

drvinceknight commented:
> OK. __eq__ was still there when I was commenting. I would vote to add it back in, but it can be another PR.

Cool. Once this is in I'll open a couple of issues:

1. To refactor the tests;
2. To add an `__eq__` method in to the players and refactor some of the base tests;
3. Once 1. is complete to refactor the base tests to remove anything that is not needed anymore.

> Do we want to change expected_outcomes to expected_actions or something else more inline with the terminology elsewhere in the library?

Good call. I've pushed that.

drvinceknight commented:
@marcharper  in bc398f2 I've gone a step further and removed `first_play_test` and `second_play_test` from the docs and `test_tit_for_tat, test_rand`. (I also added a couple more `versus_tests` in `test_tit_for_tat` to make sure all states were being tested.)

My thinking is that we can test those properties using matches and it moves us away from the potential history mistmatches that can still occur with `second_play_test`. 

Can always revert if we don't think it's a good idea.

marcharper commented:
I'd prefer to keep those tests -- the warnings for those are suppressed already. Or -- let's take that discussion to a new issue / PR?

drvinceknight commented:
> I'd prefer to keep those tests -- the warnings for those are suppressed already. Or -- let's take that discussion to a new issue / PR?

Fine by me, I've reverted the commit.

meatballs commented:
This one looks ready to roll to me. Anyone object?

drvinceknight commented:
> This one looks ready to roll to me. Anyone object?

Not from me :)
